edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.DiagnosticsComponent/MetricsWorker.cs

```diff
@@ -0,0 +1,125 @@
+
+        public MetricsWorker(IMetricsScraper scraper, IMetricsStorage storage, IMetricsUpload uploader, ISystemTime systemTime = null)
+        {
+            this.scraper = Preconditions.CheckNotNull(scraper, nameof(scraper));
+            this.storage = Preconditions.CheckNotNull(storage, nameof(storage));
+            this.uploader = Preconditions.CheckNotNull(uploader, nameof(uploader));
+            this.systemTime = systemTime ?? SystemTime.Instance;
+        }
+
+        public void Start(TimeSpan scrapingInterval, TimeSpan uploadInterval)
+        {
+            this.scrape = new PeriodicTask(this.Scrape, scrapingInterval, scrapingInterval, Log, "Metrics Scrape");
+            this.upload = new PeriodicTask(this.Upload, uploadInterval, uploadInterval, Log, "Metrics Upload");
+        }
+
+        async Task Scrape(CancellationToken cancellationToken)
+        {
+            using (await this.scrapeUploadLock.LockAsync(cancellationToken))
+            {
+                Log.LogInformation("Scraping Metrics");
+                List<Metric> metricsToPersist = new List<Metric>();
+                int numScrapedMetrics = 0;
+                foreach (var scrapedMetric in await this.scraper.ScrapeEndpointAsync(cancellationToken))
+                {
+                    numScrapedMetrics++;
+                    // Get the previous scrape for this metric
+                    if (this.metrics.TryGetValue(scrapedMetric.GetMetricKey(), out Metric oldMetric))
+                    {
+                        // If the metric is unchanged, do nothing
+                        if (oldMetric.Value.Equals(scrapedMetric.Value))
+                        {
+                            continue;
+                        }
+
+                        // if the metric has changed, write the previous metric to disk
+                        metricsToPersist.Add(oldMetric);
+                    }
+
+                    // if new metric or metric changed, save to local buffer
+                    this.metrics[scrapedMetric.GetMetricKey()] = scrapedMetric;
+                }
+
+                Log.LogInformation($"Scraped {numScrapedMetrics} Metrics");
+
+                if (metricsToPersist.Count != 0)
+                {
+                    Log.LogInformation("Storing Metrics");
+                    this.storage.WriteData(Newtonsoft.Json.JsonConvert.SerializeObject(metricsToPersist));
+                    Log.LogInformation("Stored Metrics");
+                }
+            }
+        }
+
+        async Task Upload(CancellationToken cancellationToken)
+        {
+            using (await this.scrapeUploadLock.LockAsync(cancellationToken))
+            {
+                Log.LogInformation("Uploading Metrics");
+                await this.uploader.UploadAsync(this.GetMetricsToUpload(this.lastUploadTime), cancellationToken);
+                Log.LogInformation("Uploaded Metrics");
+
+                this.storage.RemoveOldEntries(this.lastUploadTime);
```

Comment:
In 1st time run, lastUploadTime is datetime.minvalue; therefore GetMetricsToUpload will get everything from storage.  Then after upload, it call RemoveOldEntries with lastUploadTime; this will not remove anything since lastUploadTime is datetime.minvalue.  Do I understand correctly?

Another question, this.metrics is reset; as I asked before, this.metrics will have value haven't persisted to storage if those values are new and doesn't change.  Therefore metric will be lost; do i understand correctly?

Reply:
Yes, that is a mistake. It should remove entries older than the new lastUploadTime. I added a variable to store the time before the upload, and use that to remove entries and set lastUploadTime.